Feature #7756: Allow server to supply a reason to kicked/banned clients

bjarnithor99 opened the pull request:
This commit adds the missing feature of allowing the server owner to provide a reason for kicking/banning a client. The implementation extends the network protocol by adding a new network action called NETWORK_ACTION_KICKED that is capable of having an error string, unlike the other network error packages. Additionally, the kick function broadcasts a message to all clients about the kicked client and the reason for the kick.

However, for now, when banning a client, the message broadcasted says "kicked" instead of "banned".

Closes #7756

ldpl commented:
Mb I missed something but without actual GUI this looks pretty pointless as server already can send player a message before kicking and this patch doesn't seem to make it any more noticeable.

FLHerne commented:
@ldpl It looks like the patch shows the message in a popup error window (like a network disconnect).

That makes it more noticeable, but more importantly the message remains visible after being kicked; an in-game chat message sent at the same time would never be seen.

LordAro commented on src/network/network_client.cpp:
```diff
@@ -687,8 +687,17 @@ NetworkRecvStatus ClientNetworkGameSocketHandler::Receive_SERVER_ERROR(Packet *p
-	ShowErrorMessage(err, INVALID_STRING_ID, WL_CRITICAL);
+	if (error == NETWORK_ERROR_KICKED) {
+		char msg[255];
+		p->Recv_string(msg, sizeof(msg));
```
how does this work when the message was not provided on the other end?

bjarnithor99 replied:
By coincidence, there was a null byte at msg[0] when I tested my code, due to garbage on the stack.
I have added a check if there is an extra byte in the packet. If so, I assume that it is the kick message. I do this because the packet interface does not seem to have a function to check for the existence of a string.

LordAro commented on src/console_cmds.cpp:
```diff
@@ -515,28 +515,36 @@ static bool ConKickOrBan(const char *argv, bool ban)
 
-	return ConKickOrBan(argv[1], false);
+	/* Reason for kicking supplied */
+	return ConKickOrBan(argv[1], false, argv[2]);
```
I don't think anything "bad" would happen as a result, but if the reason given is longer than 255 chars, it'll be silently truncated. At the very least excess information is sent. Would be good to error message at this level rather than silently truncate

bjarnithor99 replied:
I've added a check to prevent the admin from sending a truncated message. Instead it will complain that the kick message is longer than 254 characters.

LordAro commented on src/network/network_client.cpp:
```diff
@@ -687,8 +687,17 @@ NetworkRecvStatus ClientNetworkGameSocketHandler::Receive_SERVER_ERROR(Packet *p
+		IConsolePrint(CC_DEFAULT, msg);
+		SetDParamStr(0, msg);
+		if (msg[0] == '\0') ShowErrorMessage(err, INVALID_STRING_ID, WL_CRITICAL);
+		else ShowErrorMessage(err, STR_NETWORK_ERROR_KICK_MESSAGE, WL_CRITICAL);
```
if with an else always requires braces

LordAro commented on src/network/network_server.cpp:
```diff
@@ -412,12 +412,13 @@ NetworkRecvStatus ServerNetworkGameSocketHandler::SendCompanyInfo()
  * @param error The error to disconnect for.
  */
-NetworkRecvStatus ServerNetworkGameSocketHandler::SendError(NetworkErrorCode error)
+NetworkRecvStatus ServerNetworkGameSocketHandler::SendError(NetworkErrorCode error, const char *reason)
```
doc comment needs updating + in several other places

LordAro reviewed (changes requested):
I like it

LordAro commented on src/console_cmds.cpp:
```diff
@@ -469,7 +469,7 @@ DEF_CONSOLE_CMD(ConClearBuffer)
  **********************************/
 
-static bool ConKickOrBan(const char *argv, bool ban)
+static bool ConKickOrBan(const char *argv, bool ban, const char *reason = nullptr)
```
I wonder whether this function (and the others) need to have the default parameters. Would be more consistent to explicitly pass nullptr if no reason, imo

bjarnithor99 replied:
I agree that the functions related to kicking/banning a client should explicitly pass nullptr if no kick reason is supplied. However, the function SendError is used in more cases than just kicking a client, all of which don't support printing a message at the moment. So I'm not sure that removing the default parameter and explicitly passing nullptr every time that SendError is called would increase the readability of the code. What do you think?

LordAro replied:
Yeah, that makes sense. Didn't notice that that also applied to SendError

bjarnithor99 committed:
Feature #7756: Allow server to supply a reason to kicked/banned clients

    This commit adds the missing feature of allowing the server owner to
    provide a reason for kicking/banning a client, which the client sees in
    a pop-up window after being kicked. The implementation extends the
    network protocol by adding a new network action called
    NETWORK_ACTION_KICKED that is capable of having an error string, unlike
    the other network error packages.  Additionally, the kick function
    broadcasts a message to all clients about the kicked client and the
    reason for the kick.

duckfullstop commented:
Closes #7756 when merged (thanks for implementing this!)

nielsmh reviewed (approved):
Looks fine to me.

YorVeX commented:
Nice change, appreciated!

But for the love of of God, why would you put the new NetworkAction enum value somewhere at the beginning of the enum in src/network/network_type.h - shifting all following packet numbers around for the admin interface, effectively killing all existing implementations (that otherwise would still work and only stumble about the new KICK packet if not updated)?

Good thing I found this so I knew what to look for: https://www.reddit.com/r/openttd/comments/fukwjq/server_1_and_3_version_1100_update_information_or/fmfhmey/

FLHerne commented:
@YorVeX This was noted already (#8060), and the new value has been moved to the end for 1.10.1.

Of course that means anything updated for 1.10.0 will need changing again; decision was that re-breaking some clients that are currently maintained and likely to be fixed was better than a permanent incompatibility.

See #8060 and #8061.

YorVeX commented:
Ah, sorry, my bad! Good decision there, too. Thanks for the heads-up regarding 1.10.1 ;-)
